tmt/package_managers/apt.py

```diff
@@ -0,0 +1,93 @@
+from typing import Optional
+
+import tmt.package_managers
+import tmt.utils
+from tmt.package_managers import Installable, Options, escape_packages, provides_package_manager
+from tmt.utils import Command, CommandOutput, ShellScript
+
+
+@provides_package_manager('apt')
+class Apt(tmt.package_managers.PackageManager):
+    probe_command = Command('apt', '--version')
+
+    install_command = Command('install')
+
+    def prepare_command(self) -> tuple[Command, Command]:
+        """ Prepare installation command for rpm-ostree"""
+
+        command = Command()
+        options = Command('-y')
+
+        if self.guest.facts.is_superuser is False:
+            command += Command('sudo')
+
+        command += Command('apt')
+
+        return (command, options)
+
+    def _build_presence_script(self, *packages: Installable) -> ShellScript:
+        return ShellScript(
+            f'dpkg-query --show {" ".join(escape_packages(*packages))}'
+            )
+
+    def check_presence(self, *packages: Installable) -> CommandOutput:
+        return self.guest.execute(self._build_presence_script(*packages))
+
+    def _extra_options(self, options: Options) -> Command:
+        extra_options = Command()
+
+        if options.skip_missing:
+            extra_options += Command('--ignore-missing')
+
+        return extra_options
+
+    def install(
```

Comment:
How is it making sure `apt update` is run before these commands? I think `homebrew` has similar manual update requirement.

Reply:
It's not run by tmt, just like `dnf update` is not run. I'm not sure it's documented, but it's a silent assumption, IIUIC.

Reply:
But `apt update` is different than `apt upgrade`. The former has to be run before you can do any `apt install`

Reply:
Good to know! I'm not sure tmt should run it, but since tmt now has at leats one essential requirement, `/usr/bin/flock`, maybe it will have to :/

Reply:
@LecrisUT shall we resolve this particular thread for now?

Reply:
Hm, interesting, a quick check in a fresh container:

```
# apt install tree
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package tree
```

The package installation just fails. The following simple check also fail:

```
tmt run -vvv provision -h container -i ubuntu prepare -h install -p tree login finish
...
        package: 1 package requested
            tree
            cmd: dpkg-query --show tree || apt install -y  tree
            err: dpkg-query: no packages found matching tree
            err: 
            err: WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
            err: 
            out: Reading package lists...
            out: Building dependency tree...
            out: Reading state information...
            err: E: Unable to locate package tree
        fail: Command 'dpkg-query --show tree || apt install -y  tree' returned 100.
```

So, this seems like the very basic functionality is not working for `ubuntu` images.

Reply:
@psss yes, that's a limitation of `apt`, as discussed above - unlike `yum` or `dnf`, it does not seem to pull in package info automagically. Adding the `apt update` phase unblocks you: `tmt run -vvv provision -h container -i ubuntu prepare --insert -h shell --script 'apt update' --order 1 prepare --insert -h install -p tree login finish`

It's not perfect, but it's clear why it's needed, user can do it, and once we agree how - if - tmt should do it, we can resolve it.

Reply:
Also the reason why I would be in favor of deferring this implementation is that there are other environments that could need preparation steps, e.g. `homebrew`. It would be easier to know how to tackle it, when we try to implement the other package managers as well

Reply:
@LecrisUT which implementation? :) Of the “before installing anything” phase, or this PR with package manager refactoring?

Reply:
"before installing anything" phase. :+1: from me on this PR as far as I've looked into.

Reply:
Ok, let's cover that in a separate pull request.